# Задание теме «Коллаборативная фильтрация»
Преподаватель: Наталья Баданина

ПАКЕТ SURPRISE

    используйте данные MovieLens 1M
    можно использовать любые модели из пакета
    получите RMSE на тестовом сете 0.87 и ниже

Комментарий преподавателя :
В ДЗ на датасет 1М может не хватить RAM. Можно сделать на 100K. Качество RMSE предлагаю считать на основе CrossValidation (5 фолдов), а не отложенном датасете.


In [3]:
from surprise import KNNWithMeans, KNNBasic, SVD, SVDpp
from surprise import Dataset
from surprise import accuracy
from surprise import Reader
from surprise.model_selection import train_test_split, cross_validate

import numpy as np
import pandas as pd

In [5]:
reader = Reader(rating_scale=(0.5, 5.0))
data = Dataset.load_builtin(name=u'ml-1m',)

In [24]:
data.raw_ratings

[('1', '1193', 5.0, '978300760'),
 ('1', '661', 3.0, '978302109'),
 ('1', '914', 3.0, '978301968'),
 ('1', '3408', 4.0, '978300275'),
 ('1', '2355', 5.0, '978824291'),
 ('1', '1197', 3.0, '978302268'),
 ('1', '1287', 5.0, '978302039'),
 ('1', '2804', 5.0, '978300719'),
 ('1', '594', 4.0, '978302268'),
 ('1', '919', 4.0, '978301368'),
 ('1', '595', 5.0, '978824268'),
 ('1', '938', 4.0, '978301752'),
 ('1', '2398', 4.0, '978302281'),
 ('1', '2918', 4.0, '978302124'),
 ('1', '1035', 5.0, '978301753'),
 ('1', '2791', 4.0, '978302188'),
 ('1', '2687', 3.0, '978824268'),
 ('1', '2018', 4.0, '978301777'),
 ('1', '3105', 5.0, '978301713'),
 ('1', '2797', 4.0, '978302039'),
 ('1', '2321', 3.0, '978302205'),
 ('1', '720', 3.0, '978300760'),
 ('1', '1270', 5.0, '978300055'),
 ('1', '527', 5.0, '978824195'),
 ('1', '2340', 3.0, '978300103'),
 ('1', '48', 5.0, '978824351'),
 ('1', '1097', 4.0, '978301953'),
 ('1', '1721', 4.0, '978300055'),
 ('1', '1545', 4.0, '978824139'),
 ('1', '745', 3.0, '9788

In [18]:
dir(data)

['__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 'build_full_trainset',
 'construct_testset',
 'construct_trainset',
 'has_been_split',
 'load_builtin',
 'load_from_df',
 'load_from_file',
 'load_from_folds',
 'ratings_file',
 'raw_ratings',
 'read_ratings',
 'reader']

# User based

In [6]:
algo = KNNWithMeans(k=50, min_k=2, sim_options={
    # 'name': 'cosine',
   'name': 'pearson_baseline',
   'user_based': True})
cross_validate(algo, data, measures=['RMSE'], cv=5, verbose=True)

Estimating biases using als...
Computing the pearson_baseline similarity matrix...
Done computing similarity matrix.
Estimating biases using als...
Computing the pearson_baseline similarity matrix...
Done computing similarity matrix.
Estimating biases using als...
Computing the pearson_baseline similarity matrix...
Done computing similarity matrix.
Estimating biases using als...
Computing the pearson_baseline similarity matrix...
Done computing similarity matrix.
Estimating biases using als...
Computing the pearson_baseline similarity matrix...
Done computing similarity matrix.
Evaluating RMSE of algorithm KNNWithMeans on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.8868  0.8914  0.8879  0.8888  0.8884  0.8887  0.0015  
Fit time          51.48   56.91   62.95   59.56   60.15   58.21   3.87    
Test time         112.09  123.53  131.98  114.15  115.01  119.35  7.43    


{'test_rmse': array([0.88683935, 0.8913875 , 0.8879103 , 0.88877478, 0.88843247]),
 'fit_time': (51.48081350326538,
  56.90520191192627,
  62.94666814804077,
  59.56003928184509,
  60.15226936340332),
 'test_time': (112.08706736564636,
  123.5288200378418,
  131.98305344581604,
  114.14522790908813,
  115.01060509681702)}

k=40: [0.88860119, 0.88761584, 0.89002566, 0.88949598, 0.88955333]

k=60, cosine: [0.93238378, 0.93270743, 0.93426535, 0.9351889 , 0.93508562]

k=30: [0.89139475, 0.89240682, 0.88662418, 0.88980697, 0.89030965]

min_k=3: [0.89003918, 0.8844171 , 0.88931215, 0.88974726, 0.89040256]

min_k=10: [0.88860119, 0.88761584, 0.89002566, 0.88949598, 0.88955333]

min_k=100 [1.03495605, 1.03382424, 1.038206  , 1.03493646, 1.03548635]

min_k=50 [0.9059272 , 0.90859272, 0.9068979 , 0.90715633, 0.90972776]

min_k=25 [0.89707765, 0.89257108, 0.89409967, 0.89450281, 0.89648078]

min_k=2 [0.89136969, 0.88825471, 0.88662049, 0.8890905 , 0.88854079]

# Item based

In [52]:
algo = KNNWithMeans(k=50, sim_options={'name': 'pearson_baseline',
                                       'user_based': False})
cross_validate(algo, data, measures=['RMSE'], cv=5, verbose=True)

Estimating biases using als...
Computing the pearson_baseline similarity matrix...
Done computing similarity matrix.
Estimating biases using als...
Computing the pearson_baseline similarity matrix...
Done computing similarity matrix.
Estimating biases using als...
Computing the pearson_baseline similarity matrix...
Done computing similarity matrix.
Estimating biases using als...
Computing the pearson_baseline similarity matrix...
Done computing similarity matrix.
Estimating biases using als...
Computing the pearson_baseline similarity matrix...
Done computing similarity matrix.
Evaluating RMSE of algorithm KNNWithMeans on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.8580  0.8602  0.8598  0.8591  0.8583  0.8591  0.0008  
Fit time          21.15   20.92   20.74   21.28   21.36   21.09   0.23    
Test time         51.20   52.10   51.89   56.60   58.42   54.04   2.91    


{'test_rmse': array([0.85797375, 0.8601594 , 0.85976001, 0.8591499 , 0.85825068]),
 'fit_time': (21.146010398864746,
  20.920363187789917,
  20.738446474075317,
  21.277466297149658,
  21.35996174812317),
 'test_time': (51.195528984069824,
  52.09663701057434,
  51.886112213134766,
  56.60085082054138,
  58.41980957984924)}

In [53]:
np.mean([0.85797375, 0.8601594 , 0.85976001, 0.8591499 , 0.85825068])

0.859058748

# SVD

In [48]:
cross_validate(SVD(n_factors=120, n_epochs=20),
               data, measures=['RMSE'], cv=5, verbose=True)

Evaluating RMSE of algorithm SVD on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.8748  0.8739  0.8744  0.8787  0.8715  0.8747  0.0023  
Fit time          42.16   42.06   41.80   41.60   41.49   41.82   0.26    
Test time         2.18    2.13    2.48    2.14    2.46    2.28    0.16    


{'test_rmse': array([0.87483914, 0.87391942, 0.87441464, 0.8786952 , 0.87151681]),
 'fit_time': (42.16025114059448,
  42.056405544281006,
  41.79950571060181,
  41.59631133079529,
  41.493852615356445),
 'test_time': (2.178112030029297,
  2.1311662197113037,
  2.4797701835632324,
  2.1426312923431396,
  2.4562020301818848)}

In [35]:
# cross_validate(SVDpp(n_factors=120, n_epochs=20),
#                data, measures=['RMSE'], cv=5, verbose=True)

1. Лучше с точки зрения качества работает item-based подход, это объясняется тем, что здесь пользователей больше, чем объектов - "1 million ratings from 6000 users on 4000 movies"
2. Item-based работает гораздо быстрее
3. RMSE - 0.859058748